src/relax/transform/remove_unused_parameters.cc

```diff
@@ -0,0 +1,239 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <tvm/relax/analysis.h>
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/transform.h>
+#include <tvm/relax/utils.h>
+
+#include <algorithm>
+#include <optional>
+#include <tuple>
+
+namespace tvm {
+namespace relax {
+
+namespace {
+
+template <typename T>
+using PSet = std::unordered_set<T, ObjectPtrHash, ObjectPtrEqual>;
+
+template <typename T, typename U>
+using PMap = std::unordered_map<T, U, ObjectPtrHash, ObjectPtrEqual>;
+
+struct CalleeAnalysis {
+  // The updated function
+  Function func;
+
+  // A mutator that updates calls at the call site.
+  std::function<Array<Expr>(Array<Expr>)> arg_updater;
+};
+
+std::optional<CalleeAnalysis> AnalyzeCallee(Function func) {
+  bool is_exposed = func->attrs.GetAttr<String>(tvm::attr::kGlobalSymbol).defined();
+  if (is_exposed) return std::nullopt;
+
+  auto free_relax_vars = [&]() -> PSet<Var> {
+    auto array_free_vars = FreeVars(func->body);
+    return {array_free_vars.begin(), array_free_vars.end()};
+  }();
+
+  std::vector<bool> parameter_mask;
+  parameter_mask.reserve(func->params.size());
+
+  Array<Var> params;
+  for (const auto& param : func->params) {
+    bool is_used = free_relax_vars.count(param);
+    parameter_mask.push_back(is_used);
+    if (is_used) {
+      params.push_back(param);
+    }
+  }
+
+  if (func->params.size() == params.size()) {
+    // Early bail-out for the common case where the function uses all
+    // of its parameters.
+    return std::nullopt;
+  }
+
+  // Even if a parameter is unused, it may provide definitions for
+  // symbolic variables.  We still want to remove the relax variable
+  // to reduce computational steps in the parent, but we need to
+  // provide the symbolic variables the other steps.
+  auto defined_tir_params = [&]() -> PSet<tir::Var> {
+    auto param_sinfo =
+        TupleStructInfo(params.Map([](const auto& var) { return GetStructInfo(var); }));
+    auto arr = DefinableTIRVarsInStructInfo(param_sinfo);
+    return {arr.begin(), arr.end()};
+  }();
+
+  // Use an array to define the order of the symbolic variables
+  Array<tir::Var> free_tir_vars;
+  for (const auto& tir_var : FreeSymbolicVars(func->body)) {
+    if (!defined_tir_params.count(tir_var)) {
+      free_tir_vars.push_back(tir_var);
+    }
+  }
+
+  for (const auto& tir_var : free_tir_vars) {
+    Var relax_var("param_" + tir_var->name_hint, PrimStructInfo(tir_var));
+    params.push_back(relax_var);
+  }
+
+  FuncStructInfo new_sinfo(params.Map([](const auto& var) { return GetStructInfo(var); }),
+                           func->ret_struct_info,
+                           Downcast<FuncStructInfo>(func->struct_info_)->purity);
+
+  auto arg_updater = [parameter_mask, old_relax_params = func->params,
```

Comment:
>old_relax_params = func->params

I did not know about this syntax, very cool

Reply:
Thank you.  This one was introduced in C++14, specifically to allow a lambda function to hold a move-only type.  I like using it to reduce the amount of state held by the lambda.